include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+	// optional: (useful for synchronized vector/map operations)
+	void applyRW(std::function<void(ParameterT& value)>);
+	void applyRO(std::function<void(const ParameterT& value)>);
+
+	// Copy-set/Copy-get value using read/write lock internally
+	ParameterT& operator=(const ParameterT& value);
+	ParameterT operator() const;  // or:  'operator ParameterT() const;'
+
+	// Validity Callbacks (could also use rcl_interfaces::msg::SetParametersResult like in node_parameters.h).
+	// Callbacks are stored in ordered hash map, could also possibly return hash for deleting later on.
+	// Alternatively, return Parameter<ParameterT>& for method chaining
+	using ParameterValidityCallback = std::function<bool(const ParameterT& value, std::string& message)>;
+	Parameter<ParameterT>& warnIf(ParameterValidityCallback warn_if_callback);
+	Parameter<ParameterT>& errorIf(ParameterValidityCallback error_if_callback);  // or rejectIf()/failIf()..
+	// Predefined generic (type-specific?) conditions, i.e. 'NO_OVERRIDE', 'EMPTY', 'NOT_NORMALIZED'
+	enum ParameterCondition
+	{
+		...
+	};
+	Parameter<ParameterT>& warnIf(ParameterCondition condition);
+	Parameter<ParameterT>& errorIf(ParameterCondition condition);
+
+	// Callback for reacting to simple parameter changes (read-locked)
+	// NOTE: this should be called after validity callbacks so that the parameter status can be evaluated
+	using ParameterChangedCallback = std::function<void(const Parameter& parameter)>;
+	void onChanged(ParameterChangedCallback callback);
+
+protected:
+	// Recursively declare parameters and populate child_params_, the actual node parameters (leafs) are
+	// collected and returned so that they can be maintained by NodeConfig. Should call declareChildParam()
+	// internally. NodeParameter is Parameter<rclcpp::Parameter> and needs to be forward declared..
+	virtual std::map<std::string, std::shared_ptr<NodeParameter>> declare();
+
+	// Constructor used for child parameters in the tree that share the same data, so basically all parameters
+	// that are neither root nor leaf.
+	// NOTE: not sure if this is really a good solution or if we should simply copy the data and synchronize
+	// using child/parent callbacks only
+	Parameter(const std::string& name, ParameterT& value, std::shared_mutex parent_mutex);
+
+private:
+	// Initialize child parameter and call declare() function recursively
+	template <ChildParameterT>
+	void declareChildParam(const std::string, std::map<std::string, std::shared_ptr<NodeParameter>>& node_parameters);
+
+	// Internal value protected by mutex_. Alternatively use a struct ParameterValue that manages
+	// locked acces and use that for internal data (this would allow sharing references between child
+	// parameters)
+	ParameterT value_;
+	mutable std::shared_mutex mutex_;
+
+	std::map<std::string, std::shared_ptr<Parameter>> child_params_;
+
+	const std::string name_;
+	std::string description_;
+	std::string additional_constraints_;
```

Comment:
LOL, I don't like it either, [it's just the field name of the description message](https://github.com/ros2/rcl_interfaces/blob/master/rcl_interfaces/msg/ParameterDescriptor.msg#L19). Since that one only applies to basic parameter types, we could also just rename this field to `type_specification`, `unit`, or simply `comment` (or just leave it). The reason I added this is because it would be nice to add this as inline-comment when generating yaml dumps.

Reply:
The name makes more sense in the message though because the other *modeled* constraints are listed below (ranges, etc.).
I don't think you directly model these other aspects here. I guess they are modeled in the underlying rclcpp API?

Reply:
correct! This field wasn't really thought through. Basically, you should be able to specify these for supported rclcpp parameters but in the context of more generic types this doesn't make a lot of sense. I'm not even sure if we really want to model the rclcpp constraints at all if they basically act the same way as the callbacks in the end.